### ISSUE TITLE
Is there an option to delete old (tagged) docker image if they are not associated to container

### ISSUE THREAD

**Issue by iamrommel** [githubsaturn/captainduckduck#234]
I'm on DO with only 60GB of space, and I do build regularly for my app. After 50+ builds i was running out of space already. I did some investigation and found out that there are lot of reclaimable space by running 
`docker system df`. I also ran the command `docker system df -v` and found out there are several docker images that are not associated with my active container. The docker images are the the previous builds (version) of my app. I don't need those at all, because i can build the app anytime from any tagged version on my git. In short i manage my app version from git. 

I feel that there is a use case for retaining that docker images, for let say reverting to previous version, but i guess CDD has still no option to revert to previous version.

Do you think we can add a setting on how many docker images should be retain (and setting to 0 means always the latest). These would reduce the space consume of my app on every build.

I got the clean up information from this guy https://medium.com/@ruwanka/tidy-and-sparkling-docker-environment-with-new-docker-cleanup-commands-19277f48fd89

**Comment by githubsaturn** [comment]
Disk clean up is in the docs:
https://github.com/githubsaturn/captainduckduck/wiki/Disk-Clean-Up

However, there is [an issue](https://github.com/moby/moby/issues/36295) with `docker system prune`. I am waiting for Docker people to resolve the issue so this can be part of CaptainDuckDuck. But at this point, it causes confusion. As upon running `docker system prune` you'll have to re-deploy all your apps. Otherwise, upon a container restart, your app will never wake up again.

Long story short, implementing a full automatic cleanup requires Docker fixing their problem. Having said this, I like this idea that you mentioned:

> Do you think we can add a setting on how many docker images should be retained (and setting to 0 means always the latest). These would reduce the space consuming by my app on every build.

**Comment by iamrommel** [comment]
It's weird, i was able to do `docker system prune -a` and all image (all tagged as "latest" on my apps) are deleted and only those that has associated container only remained. I did reboot the OS, and the container starts fine and even my app starts ok. So i guess everything is ok.

**Comment by drmrbrewer** [comment]
It's not clear to me from the UI what I'm meant to do... this is what it looks like without changing anything:

![image](https://user-images.githubusercontent.com/9009792/40049213-04fd94b4-582c-11e8-8d63-c2b0ba9510c4.png)

Why is the field blank?  What does `Fetch Old Images` mean?  When does disk cleanup occur... after every build?

**Comment by githubsaturn** [comment]
hmm... it's just a bad UI scaling - it doesn't play nicely on your screen size... For now, just change the zoom factor on your browser ( `ctrl` + `-` )

**Comment by githubsaturn** [comment]
Also if you hover over it, a tooltip will pop up explaining the feature. If you enter `2` in that box, it'll exclude the last two builds from the cleanup process - just in case you need them

**Comment by iamrommel** [comment]
> What cleanup process... is it something done automatically by captain? When? After each build? Periodically?

I feel you, it's not clear to me when i first see that.

But from my experience of using it, it's NOT automatic. Here are the steps
1. Set the number of image to retain, (this is  use as a filter for docker image tags)
2. Click fetch old image button, this will fetch images (excluding the last X (in my case 2) version of image)
3. Click "Remove image" button

Setting a value zero (0) to the "keep most recent" will fetch all image version (excluding the latest)


Here are the docker image before  delete
![image](https://user-images.githubusercontent.com/177265/40095575-5ffe3afc-58ff-11e8-9e33-4a3a6e46a09b.png)

And after delete (setting a value of 2 to the image that should be retained), here is the result
![image](https://user-images.githubusercontent.com/177265/40095673-d91ba9ec-58ff-11e8-9ce9-0430e6d5eeba.png)

It retains version 0, i'm not sure why, but no big deal for me. All those big size images are gone.

I love this feature to be automatic too like @drmrbrewer is expecting ( i guess), like maybe every build all image of the built app (in my case 'img-captain--hauler) will do automatically clean up.

**Comment by githubsaturn** [comment]
I've added a ton of details to the page that will answer the questions
https://github.com/githubsaturn/captainduckduck/commit/f038fff808930f4be85969acdf6df4f399e8c8ba#diff-946bb2a61343f751432e770f5aed8df7R189

**Comment by drmrbrewer** [comment]
Thanks, @iamrommel and @githubsaturn.  BTW @githubsaturn when the user upgrades captain, there is a reassuring popup to say that the update has started successfully, but after that the page just reloads with:

![image](https://user-images.githubusercontent.com/9009792/40107477-3f648468-58f0-11e8-8b09-bd4b440b473f.png)

... which is not reassuring.  It's understandable that the console will become unusable and will need to be reloaded, but maybe some sort of advance warning would be nice.

> I love this feature to be automatic too

Yes, IMHO this should happen automatically (for a particular app) after deploying another version of that app.  Or every 10, or every 20, for example.  That way, it is "set and forget".  (I note from the new info in 0.7.2 that this is planned: "_In future versions, this cleanup will be automatically done after each build._").

I'd also really love to have a "revert to version xxx" feature... I know it is on the roadmap but I hope it is sooner rather than later!

Another thing: when I click on "Get List" (with `2` in the "Keep" field), I get a lot of entries in the list that are tagged `latest`:

    registry.mysub.mydomain.com:996/captain/img-captain--myapp:latest

It seems to me that I shouldn't be deleting `latest`... and that it shouldn't be in the list?

**Comment by iamrommel** [comment]
> It seems to me that I shouldn't be deleting latest... and that it shouldn't be in the list?

I agree that shouldn't be at the list, but you won't be able to delete it as note say from the interface. CDD works as it's intention, this little/minor stuff just needs to be cleaned up. In the hands for veteran CDD user this interface is understandable but for the noobs, it will create traction for them. :)

I love this project, I used this on several production projects I have, it makes me easier to manage docker than typing docker command on my CLI

**Comment by drmrbrewer** [comment]
On the issue of disk cleanup, despite purging unused images, my disk space still seems gradually to be filling up. 

The following suggests that the `registry` subfolder of `/captain` is a big disk hog:

    # du -sk * | sort -nr | head -10
    16321892        registry
    740     letencrypt
    100     generated
    84      config.conf
    36      temp
    4       nginx-shared

Is there any way to clean that registry out, beyond what you've implemented?  There seem to be a lot of stuff in `/captain/registry/docker/registry/v2/blobs` ... does that relate in some way to a historical record of previous registry pushes (like git maintains a complete history that grows and grows)?  Any way of clearing it out... I'm not interested really in maintaining any sort of historical info.

**Comment by githubsaturn** [comment]
On the registry disk usage:
Think of Docker registry similar to git. It's where the entire history lives. If you can't accommodate the space though, you can use a third party registry ($$$). Also, note that you don't need docker registry at all, if you're deploying on a single node. If you are deploying on multiple node, and you are experimenting and not interested in keeping the history, you can remove the entire registry every now and then to free up space. I haven't tried this myself, by this shouldn't have any adverse effect on your running containers as the images they are using are already cached. To wipe the registry:
```


Some history behind docker registry:
When I was implementing the self-hosted Docker registry, I was hoping that I could find an opensource docker registry with proper UI. Unfortunately, the open-source ones are all buggy and non-functional. So I went with the default Docker registry container which does not have any UI.

**Comment by drmrbrewer** [comment]
P.S. I assume that, after wiping the registry as per your post, you then have to re-deploy every app again, so that it is again placed in the (now empty) registry, and so that you can again scale up by adding a new node?

**Comment by githubsaturn** [comment]
Yes, if you're saying more nodes. 

If you don't add any new nodes to the cluster, the images are already cached on the current nodes and do not require redeployment.

**Comment by drmrbrewer** [comment]
I ran those two commands.  Interestingly, disk utlisation showing in NetData did not change at all.  I assume that some process is still holding on to the disk space so that it cannot yet be freed up completely.

**Comment by githubsaturn** [comment]
Ahh... there is a typo in the command I copied for you, it shouldn't be `rm -rf /captain/registry/v2` it should be:
```
rm -rf /captain/registry/docker/registry/v2
docker service update captain-registry --force
```

**Comment by drmrbrewer** [comment]
OK, so now that I've removed `/captain/registry/v2`... do I need it back and if so, how?   Thanks :)

**Comment by githubsaturn** [comment]
there is no such a directory, it never existed.

**Comment by drmrbrewer** [comment]
Disk usage before/after:

![image](https://user-images.githubusercontent.com/9009792/40271617-6e5babee-5b97-11e8-8f95-ea9a43064399.png)

IMHO it would be really handy to include this sort of thing as part of the new Disk Cleanup tools.

**Comment by darrist** [comment]
This really needs to be **highlighted in Caprover official documentation** and **be included when you disable registry service** (maybe by choice).

My team went after

**Comment by githubsaturn** [comment]
Hi @darrist  -  feel free to add it to the docs. Docs are open source.

**Comment by tomaskikutis** [comment]
> [In future versions, this cleanup will be automatically done after each build.](https://github.com/caprover/caprover/commit/f038fff808930f4be85969acdf6df4f399e8c8ba#diff-8f35295b0005957f907bfe7c1ce75aee49c11701cb98b93b8f91c901f590c590R204)

Is this still on the roadmap? Also, is there a roadmap page somewhere? I saw it referenced many times, but couldn't find it neither on caprover.com nor here on GitHub.

**Comment by switz** [comment]
auto cleanup after builds would be really nice
